.park implementation

Gymnasiast opened the pull request:
# **This is not a general Q&A thread. Ask any questions on Discord. Questions placed here will be removed. Any issues you find should be reported as regular issues, making sure to mention they only affect the new save format.**

This is still in development, but it's easier to do proper discussion if we have a PR. Hence the "Draft PR".

Still to implement:
- [x] Entities
    - [x] Remove all unnecessary fields from SpriteBase (~~LinkedListId, next, previous, next_in_quadrant,~~ flags)
    - [x] Replace sprite_identifier with EntityType (remove sub types for MiscEntities)
    - [x] Replace EntityLists with lists (remove `next` field)
    - ~~[ ] Invert ownership of entities so that they store in the lists~~
    - ~~[ ] Make entity removal/addition happen at end of tick~~
    - ~~[ ] Replace EntityLists with vectors created on load (remove `next` field)~~
    - [x] Rework checksum to compute by entity type
    - [x] Rework replays to save by entity type
    - [x] Rework s6 save test to use GameState
    - ~~[ ] Investigate best way to hold all Entity types~~
    - ~~[ ] Move sprite flashing to ui project~~
    - [x] Remove all for loops over entities that use GetEntity (except where required)
    - [x] split guests and staff
- [x] Research list
- [x] Cheats (implemented by ZehMatt)
- [x] Fix crash when raising object limits (#10881)
- [x] Invisibility flag instead of corrupted elements
- [x] Revert RidesBeenOn hack
- [x] Remove `gScenarioExpansionPacks` (no longer used)
- [x] Export to .park in crash handling
- [x] Export to .parkrep in crash handling
- [x] Handle packing of objects (dat, json, parkobj)
- [x] Update replay to use .park
- [x] Allow using new _and_ old object identifiers
- [x] Handle new object types (station, surface, edge, music)
- [x] Reserve room for tile element owner
- [x] Fix scenery window to handle more scenery groups
  - [x] Use full width of window to show list items
  - [x] Fix incorrect title
- [x] Send park instead of sv6 over network (cheats now in .park data)
- [x] Splitting paths into path surfaces and path railings
  - [x] Update tile inspector
  - [x] Update plugin framework
  - [x] Use correct yellow bamboo supports
  - [x] Footpath surface objects
    - [x] JSON files
    - [x] Loading code in OpenRCT2
  - [x] Make old-style/legacy paths work as well.
  - [x] Paths without slope railings should not have railings underground
  - [x] Support colours: grey for space supports, yellow for bamboo supports
  - [x] Preview images for path surface and path supports
  - [x] Object Selection window:
    - [x] Preview for path surfaces and railings
    - [x] Tab images for path surfaces and railings
    - [x] Tooltips for legacy footpaths, surfaces and railings
    - [x] Window width
    - [x] Hide legacy footpaths from default view
    - [x] Require at least one regular path, one queue and one path railing selected. (And don't require selecting legacy paths.)
  - [x] Fix paths in track designs
  - [x] Reset footpath selection when loading a new game
  - [x] Path preview for yellow bamboo and grey space supports is broken (Ted)
  - [x] Loading SV6 files with custom objects is broken (Ted)
  - [x] Optionally: make the RCT1 space supports use the images from RCT2 + g1.dat in order to work for everyone (Michael)
- [x] Set researched ride types, entries and scenery automatically (build pre-researched list automatically)
- [x] Do we still need `gMapBaseZ`? Yes
- [x] Fix money32 / money64 issues
- [x] Replace `gScenarioTicks` with `gCurrentTicks`
- [x] Staff patrol areas
- [x] Fix read/write of notifications.
- [x] Call `scenario_remove_trackless_rides()` and `scenario_fix_ghosts()` again (or integrate them)
- [x] Miscellaneous tab in Scenery window is broken
- [x] Warn when trying to export JSON objects to TD6
- [x] Change guest history to real values
- [x] Scenario editor: Add pay park / pay rides
- [x] Fix mouse input on zoom x0.5 on half of the map (easy to repro on Electric Fields, left side)
- [x] Investigate string ID exhaustion
- [x] Fix park name always being Six Flags
- [x] Rename objects to `origin.type.name`.
  - [x] Rename space footpath surface to "red tarmac"
- [x] Remove loading of JSON legacy footpaths (DAT support for them should remain).
- [x] Fix raised circuit and art deco footpath railings (they should still have queue banners).
- [x] Allow railings to omit banners (needed for the invisible railings)
- [x] Include an invisible footpath surface

PRs to be merged first:
- [x] Refactor of remaining tile elements: https://github.com/OpenRCT2/OpenRCT2/pull/10866
- [x] Translate SV6 object offsets to OpenRCT2 ones #10881
- [x] Rework scenery handling #10922
- [x] Ride id (needed for more than 255 rides in a park) https://github.com/OpenRCT2/OpenRCT2/pull/14411

Crashes/faults to address:
- [x] News history gets corrupted

Limits to be raised in this PR:
- [x] 255 stations
- [x] 2047 small objects
- [x] 2047 large objects
- [x] 2047 wall objects
- [x] 2047 ride types
- [x] 255 combined path types
- [x] 255 banner objects
- [x] 255 path addition objects
- [x] 16777216 tile elements
- [x] 65535 sprites
- [x] 1000 rides in a park
- [x] 255 trains per ride (from 31) (#714)
- [x] 256 spawn points (up from 2)

Stuff to be left for another PR:
- [ ] More granular Z coordinates for tile elements (making them twice as granular allows for 100% RCT1 compatibility in this area).
- [ ] Splitting path elements and path additions

To be decided:
- [x] Map size (x and y are often saved in 8-bit unsigned variables, which will overflow)
- [x] Usage of old identifiers: https://github.com/OpenRCT2/OpenRCT2/issues/13080

Other stuff this PR tackles:
- [x] Making the Fruity Ices Stall colourable
- [x] Allow freezing a ride's ratings

Test notes:
- [x] Test by loading up an SV6, saving it to .PARK, then reloading that file.
- [x] Test placing track design (currently crashes in some cases)

Stuff for the changelog:
```
- Feature: [#10664] New save format with increased limits. (put this right on top!)
- Feature: [#714] Allow up to 255 trains per ride.
- Feature: [#2253] Path surfaces and railings can be mixed and matched, like in RCT1.
- Improved: [#10664] Invisibility status is now shown in the Tile Inspector.
```

Gymnasiast commented on src/openrct2/world/Entrance.cpp:
```diff
@@ -218,6 +218,27 @@ void fix_park_entrance_locations(void)
+            && entranceElement->GetSequenceIndex() == 0 && !entranceElement->IsGhost())
+        {
+            CoordsXYZD entrance;
+            entrance.x = it.x * 32;
```
Do a proper conversion here.

AaronVanGeffen commented:
With regards to the format itself, we should probably try to incorporate company ids in from the get go, even if we currently only support one company in-game. It would allow us to extend to different (competitive) forms of multiplayer games, eventually.

As for what this could look like, I'm thinking of Locomotion.

pizza2004 commented:
Does this mean we might finally see #681 implemented?  It feels like I've been waiting forever for the new save format!  Not that I've been in any rush, it's fascinating just to watch you guys work at all and it's not like you're getting paid or like the save format is the top priority.

Gymnasiast commented:
@pizza2004 Not yet. I want to avoid scope creep and only remove a handful of limits initially.

Gymnasiast commented:
@AaronVanGeffen I'm not familiar with how Locomotion achieves this. could you be more specific?

Also note that there isn't much room in most tile elements - the Path element being the worst. Where do you propose saving this value?

pizza2004 commented:
So is the goal here to make the new save format extensible in a way where removing limits won't break compatibility with older files?  I always assumed we'd want to determine all the new limits and such before settling on the new save format so it wouldn't need to change again in the future.

pizza2004 commented:
I imagine companies should own separate plots of land and everything on them, but alternatively you could tie it directly to rides and staff and peeps and the like.  I imagine the overall goal is to allow competitive multiplayer just however it makes the most sense in the context of Rollercoaster Tycoon.

Gymnasiast commented:
@pizza2004 Getting rid of limits involves more than just changing the save format.

In any case, this PR is not meant as a general Q&A thread, this is a discussion about getting the code ready for merging. If you have any questions, please just ask them on Gitter.

IntelOrca commented:
@AaronVanGeffen like @Gymnasiast said, with 16 bytes per tile element, some have room for company ID while others like footpath do not. Although we did suggest doing some like banners and have two tile elements to represent footpaths, the second being just for the additions (bins, benches).

This otherwise shouldn't affect the design of the save format as we would dump the tile element data out right, we have 8 bytes of spare bytes for each element at the moment.

Gymnasiast commented on src/openrct2/world/Surface.h:
```diff
@@ -28,7 +28,12 @@ enum
+    TERRAIN_ROOF_GREY,
+
+    // Technical types (not selectable)
     TERRAIN_CHECKERBOARD_INVERTED = 14,
```
Check if this needs to be hardcoded to 14.

Gymnasiast commented:
@IntelOrca @ZehMatt The build currently fails on targets that have network disabled (like Android), because the new save calls Crypt::CreateSHA1(), which is hidden behind an ifdef (DISABLE_NETWORK). What would be the best way to fix it?

janisozaur commented:
How about we move to sha3 and use dedicated library for that, at least for the no network builds? We still need authentication for network and the reason they are separated is to ease porting, which doesn't require openssl in the initial phase

IntelOrca commented:
@Gymnasiast Now that Windows uses native APIs to do the cryptography we can enable the crypt header for all builds. For non-Windows builds you will have to enable crypt or at least the hashing part in all builds. Implementing a SHA1 on all platforms should be easy enough, you can probably find a small copy and paste solution to that as well.

SHA1 is only for a checksum, if we want something more trivial to implement we can, I would not choose something more complex though. This is just a checksum, not a security measure.

janisozaur commented:
I would prefer moving on to the current preferred standard even in such case, it doesn't seem that difficult, especially when we design a new format.

That said, SHA1 (and later) implementations are dime a dozen. Quick Google yields e.g. https://github.com/983/SHA1

IntelOrca commented:
You will most likely have to add our own implementation of SHA-3. I don't think the Windows API supports that, and I really don't want to re-introduce OpenSSL just for this.

ZehMatt commented:
We could use a simpler hash such as xxHash for the save file, there is no need for security here.

Gymnasiast commented:
Report about saving as .PARK and reloading leading to a crash: https://github.com/OpenRCT2/OpenRCT2/issues/10872

Gymnasiast commented on src/openrct2-ui/windows/LoadSave.cpp:
```diff
@@ -219,13 +219,13 @@ static const char* getFilterPatternByType(const int32_t type, const bool isSave)
 
         case LOADSAVETYPE_SCENARIO:
-            return "*.sc6";
+            return "*.park";
```
```suggestion
            return isSave ? "*.park" : "*.park;.sc6;*.sc4";
```

Gymnasiast commented on src/openrct2/ride/Ride.h:
```diff
@@ -38,7 +38,7 @@ struct Staff;
 #define CUSTOMER_HISTORY_SIZE 10
 #define MAX_CARS_PER_TRAIN 255
-#define MAX_STATIONS 4
+#define MAX_STATIONS 8 // Increase to 254 once the tile elements can handle it
```
255

Umdlye commented:
I don't know if scenario stuff is supposed to be implemented yet, but just in case:

- Scenarios created in the new format don't show up in the scenario menu when placed in the `scenario` folder
- Loading the file directly doesn't bring up the objective screen like loading a .sc6 would
- The scenario editor doesn't require a peep spawn point to be set
- Scenario descriptions don't seem to save, the description becomes the one from Haunted Harbour regardless of what you enter
- Loading .park files through the "load landscape" option in the scenario editor gives the `File contains invalid data!` error

Slightly related:

- The default size of the object selection window needs to be bigger to prevent the new category buttons from overlapping the "filter" button

X123M3-256 commented:
I just tried to build this, and I get an error "array index 1048576 is past the end of the array (which contains 196608 elements)" on line 1112 of map.c. I'm guessing this has to do with the raised object limit?

Umdlye commented:
The path selection menu doesn't seem to like the new limit yet 😛 

17-32 paths works as expected 🎉 
33 paths crashes the game when you open the path selection
34-35 paths show the paths in one long line instead of a square menu
36-47 paths crashes the game when you open the path selection
48-49 paths show a little bit of a menu on the side of the screen
50 paths show an empty menu
51 paths show them all in one line again
52-53 paths show no menu at all but doesn't crash
54-55 paths show them all in one line
56-58 show no menu at all
59 paths show them all in one line
60 paths works as expected 🎉 
61 paths show them all in one line

Didn't test 62-253 yet

Finally 254-255 paths crashes the game without making a dump when you open the path selection

Again, sorry if this is too early, it's just cool to see that it partially works already!

Gymnasiast commented:
@Umdlye Most likely caused by trying to beyond the confines of `gAppropriateImageDropdownItemsPerRow`, which is only 34 items long.

Umdlye commented:
Yay! May be worth noting that trying to any save made in 5cbc5f9 crashes the game in 585e68d. (Windows 10, x64)

Gymnasiast commented:
@Umdlye Assume that every single commit in this branch breaks your old .PARK saves. The format will only be stable and backwards compatible once it's released.

James103 commented:
Will the new save format ever incorporate a save game version system (such as the one found in OpenTTD)? This should allow for better forward compatibility, especially when new features are added and limits are raised.

Gymnasiast commented:
@James103 Yes, but again: **this is not a general Q&A thread.**

X123M3-256 commented:
Is saving of the object selection implemented? I noticed when trying to load a save created yesterday (with the same version that was used to create it), that the void terrain type was not selected and the terrain rendered incorrectly as a result. Not sure if this is a known issue or a possible bug.

Umdlye commented:
![openrct2_2020-03-08_21-51-33](https://user-images.githubusercontent.com/11657101/76171026-949cba80-6187-11ea-98d8-0f5cf2d0dfb1.png)

Looks like there aren't enough page name strings to fit the increased amount of guests.

Gymnasiast commented:
The guest window will be addressed here: https://github.com/OpenRCT2/OpenRCT2/pull/10901

Dropdowns can only have 128 items (up from 64), so that makes sense: 96 stations + 31 trains + overall view = 128 items. We should perhaps increase the maximum size, or even look into replacing it with another widget.

Umdlye commented:
The guest counter appears to overflow at 32767 guests as reported by Marcel Vos on Discord.

https://cdn.discordapp.com/attachments/356546837945253888/686322163065683982/Test_Park_2020-03-08_22-18-55.png
https://cdn.discordapp.com/attachments/356546837945253888/686327824650469572/unknown.png

Gymnasiast commented:
@Umdlye Could you test if changing `{COMMA16}` to `{COMMA32}` on strings 1846, 1847 and 1848 fixes that problem? You don't need to recompile anything, just changing the strings in en-GB.txt and/or nl-NL.txt (depending on the language you're using) is enough.

Umdlye commented:
Doing that turned -32,505 into 16,744,711 😬

duncanspumpkin commented:
@gymnasiast you will need to modify the callers of that sprite to push a 32bit value onto the stack as well to fix the peep count.

Gymnasiast commented:
@duncanspumpkin Since there are more occurrences of the guest count being used (like in the Park window), is it perhaps a better idea to change `gNumGuestsInPark` to a 32-bit value?

duncanspumpkin commented:
Ah sorry thought it worked differently and was manually pushing the value on the args stack. That was incorrect. Yes you will need to change the type to uint32_t

duncanspumpkin commented on src/openrct2/world/Sprite.h:
```diff
@@ -194,9 +194,9 @@ enum
-extern uint16_t gSpriteSpatialIndex[0x10001];
+extern uint16_t gSpriteListHead[SPRITE_LIST_COUNT];
+extern uint16_t gSpriteListCount[SPRITE_LIST_COUNT];
+extern uint16_t gSpriteSpatialIndex[MAX_SPRITES + 1];
```
This is incorrect the spatial index should still be 0x10001

IntelOrca replied:
Nah, it should be `MAX_MAP_SIZE * MAX_MAP_SIZE + `

duncanspumpkin commented on src/openrct2/world/Sprite.cpp:
```diff
@@ -31,9 +31,9 @@ static rct_sprite _spriteList[MAX_SPRITES];
+#define SPATIAL_INDEX_LOCATION_NULL MAX_SPRITES
 
-uint16_t gSpriteSpatialIndex[0x10001];
+uint16_t gSpriteSpatialIndex[MAX_SPRITES + 1];
```
See below 0x10001 is correct

duncanspumpkin commented on src/openrct2/world/Sprite.cpp:
```diff
@@ -31,9 +31,9 @@ static rct_sprite _spriteList[MAX_SPRITES];
 static bool _spriteFlashingList[MAX_SPRITES];
 
-#define SPATIAL_INDEX_LOCATION_NULL 0x10000
+#define SPATIAL_INDEX_LOCATION_NULL MAX_SPRITES
```
revert

Gymnasiast commented:
@Umdlye Ok, should now work properly - could you re-test it?

Umdlye commented:
Looks like it works everywhere except the guests window:

![openrct2_2020-03-09_21-12-05](https://user-images.githubusercontent.com/11657101/76253315-c3c83000-624a-11ea-8157-13e38c89b183.png)
![openrct2_2020-03-09_21-11-53](https://user-images.githubusercontent.com/11657101/76253322-c62a8a00-624a-11ea-93c5-e3d6a08f2c9c.png)

duncanspumpkin commented:
That will be caused by stuffing the count of guests into var_492 which is 16 bit.

Gymnasiast commented:
@Umdlye One of my latest commits should fix the number of guests in the guest list. Could you test that?

Umdlye commented:
The overall guest count is now fixed :) The thoughts and actions counts are still broken though, not sure if this was meant to fix those too.
![openrct2_2020-03-20_20-57-05](https://user-images.githubusercontent.com/11657101/77201961-bd10a700-6aed-11ea-8ae8-39454fa14a1c.png)
![openrct2_2020-03-20_20-57-17](https://user-images.githubusercontent.com/11657101/77201967-c00b9780-6aed-11ea-96db-75fa9e19dbae.png)

Gymnasiast commented:
@Umdlye Thanks! I forgot about the summarised actions/thought counts.

Gymnasiast commented:
@Umdlye I have just pushed a commit to fix the thoughts and actions counts. Could you retest it?
(I appreciate your help a lot, btw!)

Umdlye commented:
Looks good!
https://user-images.githubusercontent.com/11657101/77229191-b127f200-6b8c-11ea-9d2e-a9c3ad35db80.png
https://user-images.githubusercontent.com/11657101/77229192-b2f1b580-6b8c-11ea-8f3c-21c0b6d3647e.png
(No problem, glad I could help!)

Umdlye commented:
![openrct2_2020-03-26_22-44-47](https://user-images.githubusercontent.com/11657101/77699988-a964b500-6fb3-11ea-8b43-dd982f16ae98.png)
![2020-03-26_22-45-19](https://user-images.githubusercontent.com/11657101/77699994-acf83c00-6fb3-11ea-853d-ad84f7f8c41a.png)

Umdlye commented:
Did some more testing, the dropdown disappears when there are more than 61537 guests. At 61537 the game just crashes when you try to open the dropdown menu.

James103 commented:
@Umdlye Why exactly does the dropdown disappear above (and crash at) 61537 guests? Could it be another signed/unsigned mishap (since 61537 is congruent to -3999 modulo 65536)

duncanspumpkin commented:
@James103 you'll get bad lag opening the guest list if you increase the length of the names by much more. Would be more performant to just increase the number of first names.

duncanspumpkin commented:
https://github.com/OpenRCT2/OpenRCT2/pull/11228 will be adding a little fix for balloon frame numbers to prevent them exceeding a byte in value. This can be reverted in the new save format. Balloon.cpp L50.

Fred-1044276 commented:
Any information on what the new limit of scenery groups will be? In OpenRCT2, it's currently still capped at 19.

James103 commented:
The cap of 19 scenery groups is due to interface limitations (width of scenery selection window for example). In order to increase this cap, the scenery selection UI would have to be redesigned to accommodate 20+ scenery groups (ideally at least 255).

Gymnasiast commented:
@James103 Not true. Please don't answer questions you don't know the answer to, it's only going to confuse people.

IntelOrca commented:
The scenery group limit is a limit with the SV6 save format. However if we do increase this limit, the interface will likely need tweaking to handle many more tabs. We may have to completely rethink the window design altogether.

IntelOrca commented on src/openrct2/ParkFile.cpp:
```diff
@@ -531,8 +531,8 @@ namespace OpenRCT2
-                    cs.ReadWrite(ride.price);
-                    cs.ReadWrite(ride.price_secondary);
+                    cs.ReadWrite(ride.price[0]);
+                    cs.ReadWrite(ride.price[1]);
```
This should write an array now.

Gymnasiast replied:
Like so: `cs.ReadWrite(ride.price);` ?

Also: the rebase meant that using std::begin, end and size on gNewsItems does not work any more, and I'm not sure how to fix it - do you have any idea?

IntelOrca replied:
Yes.

A shame, you would have to implement a begin and end method with iterators to sort it out. Or maybe you could return just pointers. Otherwise you have to iterate it with an index.

Gymnasiast commented:
@Ruedii Regular walls have only two images (or six if they can be put on a slope). Path railings have far more. Corners have inner railings, dead ends have shorter-than-regular railings, for example.

frutiemax commented:
I just pulled this branch, try to save the Magic Mountain scenario to a .park file and load it and the game logic freezes. I've debugged it and when I comment line 287 in GameState.cpp i.e. peep_update_all(), the game is no more freezed.

Gymnasiast commented:
I replaced corrupt elements in the last commit with an invisibility flag. I also updated the Tile Inspector to use it. It looks like this:

![afbeelding](https://user-images.githubusercontent.com/1478678/104839354-be069680-58c0-11eb-9ba3-e9cd10a6e53c.png)

Umdlye commented:
Awesome!

There is a small typo in the code by the way: `AnyToggleInvisilibity`.

Gymnasiast commented:
Yeah, I think it would be better. Actually, I actually tried it, but it seemed like the sprite was not compatible - there were dark bands around it. Of course, we could solve that by adding a fixed sprite to g2.dat.

duncanspumpkin commented on docs/save-format.md:
```diff
@@ -0,0 +1,255 @@
+  map-size:                 xy32
+  tile-elements:            tile-element[]
+
+Things:
```
Rename to Entities

Gymnasiast commented on src/openrct2/rct1/S4Importer.cpp:
```diff
@@ -1930,6 +1930,14 @@ class S4Importer final : public IParkImporter
             }));
         LoadObjects(ObjectType::ParkEntrance, std::vector<const char*>({ "PKENT1  " }));
         LoadObjects(ObjectType::Water, _waterEntry);
+        LoadObjects(
```
Should be removed - has been done in develop in the meantime.

Gymnasiast commented on src/openrct2/peep/Peep.h:
```diff
@@ -635,7 +635,7 @@ struct Peep : SpriteBase
     ride_id_t InteractionRideIndex;
     uint16_t TimeInQueue;
-    uint8_t RidesBeenOn[32];
+    ride_id_t RidesBeenOn[32];
```
Port to develop.

duncanspumpkin replied:
This is going to be a breaking change. I don't think it can be ported

Gymnasiast replied:
Hm, that's annoying. I guess there are no peep changes planned for the immediate future?

IntelOrca commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -1981,7 +1981,7 @@ static void window_ride_construction_entrance_click(rct_window* w)
     {
         gRideEntranceExitPlaceType = ENTRANCE_TYPE_RIDE_ENTRANCE;
-        gRideEntranceExitPlaceRideIndex = w->number & 0xFF;
+        gRideEntranceExitPlaceRideIndex = w->number % MAX_RIDES;
```
This doesn't make sense. The original code was being silly and using & 0xFF instead of a type cast. When this window becomes a class, I would expect a field like _rideIndex or something to be used.

Gymnasiast replied:
I wasn't sure about that, so I figured I'd play it safe until the window gets refactored.

IntelOrca replied:
Yeah, it would just cause confusion if it did wrap.

duncanspumpkin commented:
The money64 changes need a good bit of additional work to them. `gParkValueHistory` needs updating to 64. Need to sort out the formatter.

ZehMatt commented on src/openrct2/core/Crypt.OpenRCT2.cpp:
```diff
@@ -7,190 +7,95 @@
+        if (dataLen == 0)
+            return this;
+
+        auto src = reinterpret_cast<const uint64_t*>(data);
+        if (_remLen > 0)
+        {
+            // We have remainder, so fill rest of it with bytes from src
+            auto fillLen = sizeof(uint64_t) - _remLen;
+            assert(_remLen + fillLen <= sizeof(uint64_t));
+            std::memcpy(_rem + _remLen, src, fillLen);
+            src = reinterpret_cast<const uint64_t*>(reinterpret_cast<const uint8_t*>(src) + fillLen);
+            _remLen += fillLen;
+            dataLen -= fillLen;
+            ProcessRemainder();
+        }
```
Why is this needed? The loop of mine copies just the bytes into uint64_t, if its less than 8 bytes its null padded.

IntelOrca replied:
Would your code cope with:
  Update(5 bytes)
  Update(3 more bytes)

and give same result as if you did Update(all 8 bytes)?

ZehMatt replied:
No. A lot of hash algorithms that use blocks will not do this.
https://godbolt.org/z/1875K31nG

There is also really no reason to have this, this is not a cryptographically secure hash, they also can have different schemas.

IntelOrca replied:
> There is also really no reason to have this

Sorry I disagree with this. The API is so that you can stream the data into the hash algorithm. It should not give you different results depending on the size of the chunks you feed it.

ZehMatt replied:
How would you implement lets say Murmur2 like that?

IntelOrca replied:
Exactly the same way.

ZehMatt replied:
I mean okay, it doesn't matter too much. The thing that sticks out the most is reinterpret_cast, you should avoid it when possible, since we are not on C++20 yet there is no std::bit_cast so the std::memcpy will have to do.

Umdlye commented:
Early observations from 27e83dc:
* Invention list changes in the scenario editor/through the cheat are not saved properly. Only the 'pre-invented' selection is saved. The 'to invent' selection is not, and the list there is duplicated multiple times
* Loading a .park scenario as a saved game does not bring up the objective window like loading a .sc4/.sc6 would

LordMarcel commented:
The sprite limit has been raised and seems to work fine, but the staff limit is still 200. According to Gymnasiast this is an oversight and he told me to report it here. Amazing work on all this stuff by the way.

oli414 commented:
The sign limit appears to still be 200, got the message for too many banners on a 14x14 section of path with banners placed in 2 directions:
![image](https://user-images.githubusercontent.com/2348094/117516688-18278680-af9a-11eb-83f2-11ee26314bc8.png)

nottud commented:
![image](https://user-images.githubusercontent.com/1540756/117519756-d3511f00-af9c-11eb-9da7-57ab0f2ddcec.png)

Corruption on the main menu.

nottud commented:
![image](https://user-images.githubusercontent.com/1540756/117519770-ea900c80-af9c-11eb-9cf2-2ce51bf0e74b.png)

Corruption on main menu

nottud commented:
![image](https://user-images.githubusercontent.com/1540756/117519807-288d3080-af9d-11eb-9990-0904f8bab342.png)

May be issue exclusively for RCT1 scenario files?

nottud commented:
![image](https://user-images.githubusercontent.com/1540756/117519839-4e1a3a00-af9d-11eb-89f1-e4c6596d1554.png)

Seems to also occur on newly placed path too.

nottud commented:
Least one ride tested does not take advantage of the new max limit of vehicles by default. It currently does 32 but can be raised to 58 (Cannot raise above this due to max platform length limit). Possible situation is similar on other rides.

Richard-L commented:
We noticed some sprite culling difficulties with RCT1 parks. This Top Spin one was reported by https://github.com/Xtreme97-RCT.

Park link: https://www.nedesigns.com/park/152/harakiris-islands-of-adventure/

* [ ] Sprite culling
RCTLL:
![image](https://user-images.githubusercontent.com/4934209/117544965-07235780-b024-11eb-9463-0bdee5628fce.png)
OpenRCT2, v0.3.3 (417e5ae on develop):
![image](https://user-images.githubusercontent.com/4934209/117545045-6e410c00-b024-11eb-824f-5f5a5fff0997.png)
NSF build from last night:
![image](https://user-images.githubusercontent.com/4934209/117545051-7600b080-b024-11eb-97a6-c7ac7cc485fc.png)

giraty commented:
- [ ] A bit of issue with rides with wooden style base supports 
[img]https://cdn.discordapp.com/attachments/691736850980667514/840615182022738010/unknown.png[/img]

[img]https://cdn.discordapp.com/attachments/691736850980667514/840615231884361738/unknown.png[/img]
- [ ] Don't know if this is intended or not, but new type scenario has their own category in Other Parks
![image](https://user-images.githubusercontent.com/8456673/117545379-bde8fd00-b04f-11eb-8eb7-d4270b30c6b6.png)

Krutonium commented:
![Profit](https://user-images.githubusercontent.com/3945538/117552487-9ecf7880-b019-11eb-868c-0fbf01e31d80.png)
Seems to be on at least RCT1 Scenarios.

Krutonium commented:
After saving and reloading that same park; 
![Research](https://user-images.githubusercontent.com/3945538/117552597-6da37800-b01a-11eb-8f47-02780359c54a.png)

[Forest Frontiers.zip](https://github.com/OpenRCT2/OpenRCT2/files/6446503/Forest.Frontiers.zip)

Fred-1044276 commented:
- [ ] Path additions from other scenery groups (e.g. Mine Themeing, Jurassic Themeing, Roaring Twenties Themeing) do not show up in any of the default scenery groups (Trees, Shrubs and Ornaments, Gardens, Fences and Walls, Walls and Roofs, Signs and Items for Footpaths) at all:
![Metropolis 2 2021-05-08 13-50-48](https://user-images.githubusercontent.com/61885939/117553121-6e311400-b004-11eb-9743-a82493f87117.png)

loweml commented:
I've found an issue with the average cash guests come in with. I've tested it with the most recent version that doesn't use the new save format and didn't have this problem. When guests come into the park though, their initial cash is as it should be. It only seems to be an error in the Scenario options menu. Not a problem at all, at least so far, but I just thought you should know about it. 

![9594b1723416a6037b916773d8c76193](https://user-images.githubusercontent.com/82008764/117556832-1661cf80-b03b-11eb-8b7a-eadc48ca6153.png)
![e43b3af725f77da6c7a1c89a3be41ade](https://user-images.githubusercontent.com/82008764/117556833-16fa6600-b03b-11eb-8747-3b33239dd28b.png)

simaticable commented:
Woodworm Park has a locomotive instead of the tenders on the trains.
![2021-05-09](https://user-images.githubusercontent.com/83894317/117560303-a82d0500-b05a-11eb-8d81-f72ae0e0cd9a.png)

keiyakins commented:
- [ ] Incorrect scenario marked complete in scenario list on completion of game loaded from .park save. Tested with Trinity Islands and Katie's World saves. TI marked Six Flags Magic Mountain, then itself. Katie's World marked Funtopia then Aqua Park. Will test an RCT2 scenario now...

LurkerRCT commented:
That title screen thing would make sense, every scenario that can be finished on my game already is, and my scenario was a small one with a short goal so what might have happened was it tried to mark one of those and it had a higher company value than the one I played, so no name option.

Gymnasiast commented:
- [ ] Restricting scenery on RCT1 scenarios also hides items that attach to a scenery group, like rct1.wfw0.

esojka commented:
Since the elimination of the object limits is a request that is like 100 years old, you know the first thing people are going to do will be to select-all on the Object Selection screen.  A couple of issues arise in the GUI when a user does this:

- [ ] The existing filters help in ths regard, but it would be nice to have a "Select All" on the selection screens to avoid lots of clicking to add multiple items.
- [ ] Once multiple scenery groups are selected, if the number of scenery groups exceeds the width of the Add Scenery window, only as many as can be displayed are shown, with the individual objects from the other SGs placed into the "?" tab.

esojka commented:
Using the eyedropper tool in the scenery selection screen causes a crash with this build

Fred-1044276 commented:
- [ ] It is impossible to scroll through the Recent Messages window in the .park test build. The scrollbar refuses to budge.
- [ ] money64 issue in the water tool.

captnpineapple commented:
- [ ] > Using the eyedropper tool in the scenery selection screen causes a crash with this build

Adding checkbox - as I encounter this error too, and the checkbox was not on the last person's reply

AWildShinx commented:
![image](https://user-images.githubusercontent.com/67431060/118233661-679b1480-b4bc-11eb-9212-1f8f84156b53.png)
It seems generator extremely obsessed with arborginal art pieces 
Reproduction steps: 
1. Select all objects or choose arborginal art piece 
2. Generate terrain with trees
3. You should see many arborginal art pieces there

Gymnasiast commented:
@AWildShinx That also affects develop: https://github.com/OpenRCT2/OpenRCT2/issues/14155

StrangeAnomoly commented:
- [ ]  > Using the eyedropper tool in the scenery selection screen causes a crash

ridetester101 commented:
New here so not too sure of where to feedback stuff yet: just posted some related here

https://github.com/OpenRCT2/OpenRCT2/issues/14663

Two other thoughts- as paths and footpath surfaces are two different things are there any potential pitfalls here with custom paths vs footpath surfaces plus railings, similarly will sharing files be a problem as it could be a lot of new objects in one go for the other player.

Love the idea of this update- it will be worth the headaches getting it working (I'm not a coder but could think and help with logical input such as what might be good for quality of life improvements etc as handling lots of objects is great for creativity in game but a massive headache in terms of time and having to scroll through 000s of items which may or may not end up in your game, so fixing some of these things could also be a chance to add some quality of life elements such as scenery management)

One would be being able to select too many items whilst making your choices to then whittle the long list down, thus a sort of in-progress landscape editor- at this early stage it would just need to save which choices are on/off and have a state where you can't progress until you've cut any excess, eg you may have say 120 too many so you can go back later and see what you can cut - reason is with so many choices now this is a big burnout as it's kind of designed to have to do it in all one go currently which is too much, so a quality of life workaround to this would be great, as would others like being able to define your own tags/or and groups, say trains or cars or sci fi etc, but that's a bit wishful thinking for now, one for when other stuff is resolved.

Thanks for the epic work on this so far and keeping the game alive in 2021.

ridetester101 commented:
Re esojka's screenshot of blanks in scenery box I had some like that with a small mystery image and then the bottom of the long scrolling previews were totally blank- some preview images have problems so you got some blanks etc even before this limit-breaking update, it may be designers setting up preview images wrong or preview images going a bit awol or some other problem with specific scenery items and how they were set up, again before this limit-breaking there were still problems with scrolling selected items- you could pick scenery items and paths that still didn't appear in game, possibly as the scrolling windows had run out of slots/ something else capacity related. Happens in some other older games on the official releases- one, maybe locomotion, let you select more available trains than would actually appear in the vehicle selection window.

So any limit breaking also needs to match what will/won't work within any in-game limits. Plus of course working out what those limits are and how to work most efficiently within them/ if they can be worked around.

LordMarcel commented:
Two more bugs here:

Normal path without railings is invisible. Queue lines work fine.
![Screenshot_353](https://user-images.githubusercontent.com/25930360/118377890-5e02d100-b5d0-11eb-83c0-892e9ab2263c.png)

When building a coaster vertically into the ground the underground track sometimes doesn't get rendered.
![Screenshot_352](https://user-images.githubusercontent.com/25930360/118377934-a02c1280-b5d0-11eb-9c49-6ab15ea448c9.png)

Gieneq commented:
![image](https://user-images.githubusercontent.com/23253207/118452647-a74d4080-b6f6-11eb-9005-427fc2cc26af.png)
Very messy. i7-9750HF, RTX 2060 Win10

Gymnasiast commented:
@Gieneq Not a bug. You haven't linked your RCT1 install properly.

jmciver92 commented:
Im not sure if this matter but I am testing on the MacOS build.

- [ ] Screen Crash when using the Eye-Dropper Tool

ridetester101 commented:
Have run some further tests in scenario and landscape editor- does seem broken- made various attempts to use the Blackpool Pleasure Beach scenario, all ultimately failed

First related issue is replacing original files with the .park.sv6 file- this in itself is problematic as if the .park.sv6 file has issues you can't go back to the original so means a lot of multiple backing up/ repeating and the .park.sv6 files it seems are sometimes unloadable- of course it could be that they shouldn't have the double extension and just be .park depending how you've implemented the new save format, but due to issues it would be worth preserving the original file rather than replacing it if it is a non-park file.

First tried loading the park from a save to convert a saved park to a scenario- worked and loaded into scenario editor but as soon as I saved it the problems began- would not then reload anything when I tested exiting and seeing if could reload into editor (just got the blank empty map again), interestingly even after making the save again from the game and then trying to import the save game into the scenario editor I got nothing (maybe for now we also need option to save non-park games as either park format or original save format, it seems to be the park format it dislikes) so weirdly even the thing that had worked in test 1 failed in test 1b, the repeating of test 1.

Tried various things a few more times and eventually through the load landscape method got the landscape without the rides successfully, the entrances and everything seemed fine with pathing to map edge but again after saving it and exiting it did not want to reload the new park file into the editor- wanted to test if it works if you do the next steps like invention list setup but it has since refused to reload the park either as just landscape or with rides, seems the park format is causing a problem the editor can't handle.

Not sure if this being an RCT1 scenario originally is any part of the issue but it is similar to other problems I had trying a few things out with the editor, somewhere along the lines you seem to hit a point where you can't load files previously saved in the editor.

One thing I noticed is the create scenario from saved park option has no 'set up the objectives' bit- just skips to save scenario as the first 'next' option after loading park, IF you can get a landscape only to load from such a save it does have the correct next step of invention list setup, will try a few more experiments, including for rct2 parks that aren't real parks in case there are issues with either rct1 parks or real parks.

ridetester101 commented:
Scenario/landscape editor test updates.

Found even though editor refuses to reload them (sc6 files) the game will through the load saved game tab so the data it seems is in tact if can fix editor and other issues, when checking in game:

Bottom of item selection window cuts off at higher zoom (window scale) levels in options

It will load a park where guest spawn path doesn't reach edge of map but this will have a glitched out of park path area very full of peeps- still works and didn't crash but wonder if it would eventually fail. If a design queue wasn't set it randomizes one. Positive as lots of attractions= can get innovations over decades so could make for a long and satisfying game, or test players patience! Invented shops etc were getting added and available and worked. Not sure if any were blanks as only checked a few.

Some blank rides could be placed and were some of the ones with pre-built layout/ design your own layout choices, but there could still be glitches as didn't check if these blank ones worked properly. Some had multiple options for flat rides but blank ride build window options were unavailable after placing.

Not available in next point is ones showing blanks or rides with unavailable vehicle options- didn't pick up all of these which should explain my only accounting for 670 attraction types. These tend to just plonk down nothing, a triangle, or a flat ride base with no ride, usually glitched, but possibly allowing a different flat ride to be selected in the ride's options menu.

1560 available scenery etc pieces versus 3122 selected scenery etc pieces, no missing scenery groups but had only selected six, missing items were not just in misc tab though so these other tabs had some missing too

Approx 492 available attractions with 178 not available which were often glitchy if placed (=670) out of 718 selected

MORE INFO TO FOLLOW - WILL BE BACK LATER (run out of time for now)

Gymnasiast commented:
@jmciver92 That has been reported several times already. Also, Windows != PC.

Gymnasiast commented:
@ridetester101 Can you add checkboxes to your points like everyone else? Makes it easier for us to check points that have been fixed.

KevsterW21 commented:
Ive come across an issue in this version where the rides "Chase" camera doesnt seem to work anymore. Ive installed this version fresh & havent had any issues with any other game versions of this not working. 
![Openrct2 Screenshot 2021 05 17 - 03 40 47 21](https://user-images.githubusercontent.com/84298950/118722586-39069c00-b824-11eb-8b61-adab23fc0d4e.png)

Gymnasiast commented:
@KevsterW21 I think that was already reported here: https://github.com/OpenRCT2/OpenRCT2/pull/10664#issuecomment-841208899

Gymnasiast commented on src/openrct2/ParkFile.cpp:
```diff
@@ -0,0 +1,2091 @@
+
+        void ReadWriteBannersChunk(OrcaStream& os)
+        {
+            os.ReadWriteChunk(ParkFileChunkType::BANNERS, [this, &os](OrcaStream::ChunkStream& cs) {
```
`this` is unused. Clang chokes on this.

cuyama commented:
The scenery eyedropper causes a crash when I try to pick an object. As far as I can tell this is independent of the object you're trying to select, map size, and whether the current park was loaded from an SV6 or a .park.

Fred-1044276 commented:
- [ ] Attempting to start up any preexisting .sc6, .sv6, or .park file containing RCT2 Time Twister footpaths will fail (confirmed on the newest build from the "[Actions](https://github.com/OpenRCT2/OpenRCT2/actions)" hub). This is because the game is still using old object names for the RCT2 Time Twister footpath railings. The scenario I tried to open is Mythological - Cradle of Civilization 2 (Mythological Madness 2):
![Vertigo Vale 2021-05-20 17-24-33](https://user-images.githubusercontent.com/61885939/119064559-712aed80-b990-11eb-827b-44b3f915aba7.png)

Fred-1044276 commented:
- [ ] Guests' time in park stats are imported incorrectly when starting up a scenario or loading up a save game that already has guests in the park:
![Luminous Landscapes 2021-05-21 10-23-47](https://user-images.githubusercontent.com/61885939/119175993-64090f80-ba1f-11eb-88d8-6bf60c37077a.png)
![Animatronic Antics 2 2021-05-21 10-35-45](https://user-images.githubusercontent.com/61885939/119176774-5607be80-ba20-11eb-82c0-ce872f3812c1.png)

As you can see, in the top picture, Guests 1 and 12 are claimed to have spent 1,085 h 24 m and 1,085 h 23 m, respectively, in the park, which just looks wrong (the code apparently added almost 65,536 minutes to the time in park stats of all guests in the park). The picture below shows Mythological - Animatronic Film Set 2 (Animatronic Antics 2), with Guests 6537 and 8130 having time in park stats of 1,018 h 6 m and 1,016 h 16 m, respectively. If I start up the same scenarios in the current develop or release builds, the issue doesn't occur as Guests 1 and 12 have correctly spent 13 and 11 minutes in the Luminous Landscapes park, respectively, and Guests 6537 and 8130 have correctly spent 2 h 9 m and 19 m in Animatronic Antics 2, respectively.
![Luminous Landscapes 2021-05-21 10-30-33](https://user-images.githubusercontent.com/61885939/119176185-a0d50680-ba1f-11eb-8fe1-8c9b921a4d07.png)
![Animatronic Antics 2 2021-05-21 10-34-59](https://user-images.githubusercontent.com/61885939/119176781-5902af00-ba20-11eb-9629-42ec7ea0ef52.png)

- [ ] In the build scenery window, any scenery objects that appear in **two** different scenery groups, one of them usually being a default scenery group (Trees, Shrubs and Ornaments, Gardens, Fences and Walls, Walls and Roofs, or Signs and Items for Footpaths), do not appear in any of the default scenery groups if their unique scenery groups are available to use (Giant Candy and Wonderland Themeings are unaffected though). This is because all affected scenery objects' json files have a "sceneryGroup" flag which doesn't work correctly in current NSF build:
![Animatronic Antics 2 2021-05-21 10-38-33](https://user-images.githubusercontent.com/61885939/119177504-42a92300-ba21-11eb-876f-0c1a53a786be.png)
![Animatronic Antics 2 2021-05-21 10-38-38](https://user-images.githubusercontent.com/61885939/119177514-450b7d00-ba21-11eb-9ef9-3af40766b0e9.png)

On develop and release, the "sceneryGroup" flags function properly.

Apologies for having reported these issues earlier, but I wanted to report them again with some clarification this time around.

Fred-1044276 commented:
- [ ] Scenarios with park value objectives are shown with an incorrect end date on the scenario selector screen. Top pic shows a park value goal of $550,000 at the end of October, Year **11** and was taken in the current NSF test build. Bottom pic was taken in current develop build and shows the same park value goal but with the correct end date of October, Year **3**.
![Six Flags Magic Mountain 2021-05-21 11-13-58](https://user-images.githubusercontent.com/61885939/119180948-d250d080-ba25-11eb-9d82-fb9ce5315247.png)
![Six Flags Magic Mountain 2021-05-21 11-14-39](https://user-images.githubusercontent.com/61885939/119180949-d381fd80-ba25-11eb-86a1-d9b2be3fa31c.png)
This affects all scenarios with a "Park value at a given date" objective.

Porthor commented:
- [ ] Completing the park Rainbow Valley (Pulled from the Deluxe version of RCT1) checks off Haunted Harbor.

![2021-05-21 21_14_59-Start](https://user-images.githubusercontent.com/44329198/119214472-fc34e200-ba7b-11eb-89d7-4f542cd09768.png)

ZeeMajora commented:
Fruity ice cream stalls are imported with colors when importing .sc6 scenarios and .sea (rct classic) scenarios.
Tested .sc6 with ayer's rock and six flags over texas.
Tested .sea with gentle glen and haunted harbor.
Curiously the rct classic scenarios i tested still seem to have rct1's color data present for these stalls so it may be a good idea to exclude them from being imported with light blue paint as the colors are the same as they are in rct1's equivalent scenarios.
![DeepinScreenshot_openrct2_20210523214941](https://user-images.githubusercontent.com/42477864/119286022-7054be80-bc11-11eb-99ac-9361eb8cde91.png)
![DeepinScreenshot_openrct2_20210523215402](https://user-images.githubusercontent.com/42477864/119286025-721e8200-bc11-11eb-8923-8dc0724e3ae0.png)
![DeepinScreenshot_openrct2_20210523214425](https://user-images.githubusercontent.com/42477864/119285468-18698800-bc10-11eb-8a33-25283bc2110b.png)
![DeepinScreenshot_openrct2_20210523214327](https://user-images.githubusercontent.com/42477864/119285470-19021e80-bc10-11eb-9bd8-24b248ee1603.png)

Gymnasiast commented:
**To everyone: since this issue has become hard to go through, we are changing the way bugs are reported. You should now simply create an issue as you would for a regular bug, making sure to mention it is in the new-save-format branch.

AaronVanGeffen commented on src/openrct2/ride/Ride.cpp:
```diff
@@ -2326,6 +2330,7 @@ static void ride_breakdown_update(Ride* ride)
 {
     if (gCurrentTicks & 255)
```
MAX_RIDES?

ZehMatt replied:
I wanted to keep the replays functional, its also not 100% clear what this is supposed to achieve, it ticks updates 254 times leaving only the first tick out. In another test run the vehicles got stuck mid way for some reason, so its for another time.

IntelOrca replied:
I am not sure 255 is representing `MAX_RIDES` here, it is just trying to make this method run once per 256 ticks. Similar to the every 8192 ticks below.

ZehMatt replied:
It runs whenever value > 0, but I can see how one can be fooled by this line, it got me too.

ZehMatt replied:
I did not notice the return, what Ted says.

duncanspumpkin commented on src/openrct2/ride/Ride.cpp:
```diff
@@ -2078,8 +2078,12 @@ void Ride::Update()
     if (lifecycle_flags & (RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN | RIDE_LIFECYCLE_DUE_INSPECTION))
-        if (((gCurrentTicks >> 1) & 255) == static_cast<uint32_t>(id))
+    {
+        // Breakdown updates are distributed, only one ride can update the breakdown status per tick.
```
But why. That seems like a bad idea. Perhaps if we did `(gCurrentTicks & 255) == (rideId & 255)` then it does not change anything balance wise from the base game but allows for the new limits

ZehMatt replied:
I think the reason it was built like this is to avoid having news popups overwriting each other. If I wrap the ride id its possible that two or more rides breakdown each adding the news entry in the same tick where normally the distance of breakdowns is delayed by ride id distance.

Gymnasiast replied:
Can we get a resolution for this?

duncanspumpkin replied:
I still feel we should do it `(gCurrentTicks & 0xFF) == (rideId & 0xFF)` as is atm rides will breakdown much less frequently.

ZehMatt replied:
Makes sense

duncanspumpkin commented on src/openrct2/ride/Ride.cpp:
```diff
@@ -2078,8 +2078,12 @@ void Ride::Update()
     if (lifecycle_flags & (RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN | RIDE_LIFECYCLE_DUE_INSPECTION))
-        if (((gCurrentTicks >> 1) & 255) == static_cast<uint32_t>(id))
+    {
+        // Breakdown updates are distributed, only one ride can update the breakdown status per tick.
```
But why. That seems like a bad idea. Perhaps if we did `(gCurrentTicks & 255) == (rideId & 255)` then it does not change anything balance wise from the base game but allows for the new limits

Porthor commented:
- [ ] Top Spin vehicle renders on top of everything. Rides, scenery, ect.

![2021-05-26 22_09_18-OpenRCT2](https://user-images.githubusercontent.com/44329198/119770017-75fc0f80-be70-11eb-83c8-3b3c2c4309a6.png)

esojka commented:
Elevator ride is wonky in this build - Elevators require multiple stations.  If you build second/third/etc. stations and then add  entrance/exit to each, all previously built stations disappear.  Build entrance/exit on first station, then close construction window, that E/E stays.  Go back into construction mode, build subsequent E/E all of them disappear.

Gymnasiast commented on src/openrct2/object/FootpathObject.cpp:
```diff
@@ -100,6 +100,7 @@ void FootpathObject::ReadJson(IReadObjectContext* context, json_t& root)
                 { "hasSupportImages", RAILING_ENTRY_FLAG_HAS_SUPPORT_BASE_SPRITE },
                 { "hasElevatedPathImages", RAILING_ENTRY_FLAG_DRAW_PATH_OVER_SUPPORTS },
                 { "editorOnly", FOOTPATH_ENTRY_FLAG_SHOW_ONLY_IN_SCENARIO_EDITOR },
+                { "noSlopeRailings", FOOTPATH_ENTRY_FLAG_NO_SLOPE_RAILINGS },
```
Can be taken out, as we never use this for legacy footpaths.

Gymnasiast commented on src/openrct2/drawing/ImageImporter.h:
```diff
@@ -45,13 +45,17 @@ namespace OpenRCT2::Drawing
         };
 
+        ImportResult Import(
+            const Image& image, int32_t srcX, int32_t srcY, int32_t width, int32_t height, int32_t offsetX, int32_t offsetY,
```
Using ScreenCoords here might be better.

Gymnasiast commented on src/openrct2/FileClassifier.cpp:
```diff
@@ -62,6 +69,29 @@ bool TryClassifyFile(OpenRCT2::IStream* stream, ClassifiedFileInfo* result)
+    try
+    {
+        auto magic = stream->ReadValue<uint32_t>();
+        if (magic == 0x4B524150)
```
This needs a constant.

duncanspumpkin commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -7459,6 +7461,6 @@ void Guest::RemoveRideFromMemory(ride_id_t rideId)
         auto& lastEntry = Thoughts.back();
         lastEntry.type = PeepThoughtType::None;
-        lastEntry.item = PEEP_THOUGHT_ITEM_NONE;
+        lastEntry.argument = 0;
```
Should be std::numeric_limits<uint32_t>::max() to match L6800

Gymnasiast commented on src/openrct2-ui/windows/MapGen.cpp:
```diff
@@ -818,7 +817,7 @@ static void window_mapgen_simplex_mouseup(rct_window* w, rct_widgetindex widgetI
 
             mapgenSettings.height = _baseHeight;
-            mapgenSettings.water_level = _waterLevel + 2;
+            mapgenSettings.water_level = _waterLevel + MINIMUM_WATER_HEIGHT;
```
The changes in this file (and similar ones) should be backported.

Broxzier replied:
This was introduces in https://github.com/OpenRCT2/OpenRCT2/pull/14700, when raising the water height limit. This commit can safely be cherry-picked: 6a64b5a63cf24d7a9baf8f398445d341fc65c179

Gymnasiast replied:
OK, can you explain the usage `MINIMUM_WATER_HEIGHT + MAXIMUM_WATER_HEIGHT` to me? It’s in a place where I would expect it to just read `MAXIMUM_WATER_HEIGHT`.

Broxzier replied:
Good question, I was thinking the same thing yesterday. The game originally stores the water level in 5 bits, each step increases the base height by 2, and 0 means there is no water at all. This leaves us with 31 usable water levels. Since in the original games, surface element cannot have a base height under 2, a minimum water level of the same base height (value 1) would be useless. Instead of wasting this value, Chris Sawyer decided to add a minimum water level which essentially is an offset. With this offset, we gain one additional usable water level while still only using 5 bits.

The naming of `MAXIMUM_WATER_HEIGHT` is a bit confusing, as it's not the absolute height, but the maximum offset from the minimum water level.

Broxzier replied:
| Value | Water surface height |
| --- | ---
| 0 | None |
| 1 | 4 |
| 2 | 6 |
| 3 | 8 |
| ... | ... |
| 29 | 60 |
| 30 | 62 |
| 31 | 64 |

Gymnasiast replied:
I see. That sounds like a good candidate for refactoring - the offset is a detail that should only be known to the surface tile element itself. I might create an issue for that.

Gymnasiast commented on src/openrct2/world/Footpath.h:
```diff
@@ -49,6 +44,23 @@ struct rct_footpath_entry
     uint8_t flags;                        // 0x0B
     uint8_t scrolling_mode;               // 0x0C
+
+    constexpr uint32_t GetQueueImage() const
```
Parts of the footpath refactor could also be backported.

Gymnasiast commented on src/openrct2/object/ObjectRepository.cpp:
```diff
@@ -631,14 +657,16 @@ bool IsObjectCustom(const ObjectRepositoryItem* object)
     // will try to pack them into saved games.
-    auto type = object->ObjectEntry.GetType();
-    if (type > ObjectType::ScenarioText)
+    if (object->Type > ObjectType::ScenarioText)
```
Can be removed now we no longer export to SV6.

Gymnasiast commented on src/openrct2/paint/tile_element/Paint.Entrance.cpp:
```diff
@@ -14,6 +14,8 @@
 #include "../../drawing/LightFX.h"
 #include "../../interface/Viewport.h"
 #include "../../localisation/Localisation.h"
+#include "../../object/FootpathObject.h"
```
revert

ZehMatt commented on src/openrct2/actions/RideCreateAction.cpp:
```diff
@@ -297,7 +297,13 @@ GameActions::Result::Ptr RideCreateAction::Execute() const
+    ride->entrance_style = OBJECT_ENTRY_INDEX_NULL;
+    if (rtd.HasFlag(RIDE_TYPE_FLAG_HAS_ENTRANCE_EXIT))
+    {
+        ride->entrance_style = gLastEntranceStyle;
```
I think it might be better to pass this as a parameter to the game action, this global may be different on other clients, I haven't looked  where this is assigned or when.

duncanspumpkin commented on src/openrct2/interface/Window.h:
```diff
@@ -318,19 +316,6 @@ struct map_variables
     uint16_t var_488;
 };
 
-struct ride_variables
```
These should be reverted

duncanspumpkin commented on src/openrct2/actions/LargeSceneryPlaceAction.cpp:
```diff
@@ -220,6 +221,12 @@ GameActions::Result::Ptr LargeSceneryPlaceAction::Execute() const
 
     res->Position.z = maxHeight;
 
+    if (!CheckMapCapacity(sceneryEntry->tiles, totalNumTiles))
```
These changes can be reverted. Its checked in the query.

duncanspumpkin commented on src/openrct2/core/IStream.cpp:
```diff
@@ -57,6 +54,17 @@ namespace OpenRCT2
         }
     }
 
+    void IStream::WriteString(std::string_view str)
```
This can be backported.

duncanspumpkin commented on src/openrct2/core/MemoryStream.cpp:
```diff
@@ -262,6 +262,12 @@ namespace OpenRCT2
         Write<16>(buffer);
     }
 
+    void MemoryStream::Clear()
```
backport.

Gymnasiast commented on distribution/openrct2.d.ts:
```diff
@@ -517,7 +519,7 @@ declare global {
     }
 
-    type VehicleCrashIntoType = "another_vehicle" | "land" | "water";
+    type VehicleCrashIntoType = "another_vehicle"|"land"|"water";
```
Formatting changes in this file should be reverted.

Gymnasiast commented on src/openrct2-ui/windows/Ride.cpp:
```diff
@@ -781,7 +781,7 @@ static std::unique_ptr<TrackDesign> _trackDesign;
 // (Re)calculated when the ride window is opened
 struct ride_overall_view {
-    int16_t x, y, z;
+    int32_t x, y, z;
```
Can be backported.
Better yet, changed to CoordsXYZ or TileCoordsXYZ.

duncanspumpkin replied:
yeah doing that atm.

duncanspumpkin commented on src/openrct2/interface/InteractiveConsole.cpp:
```diff
@@ -1278,15 +1278,12 @@ static int32_t cc_show_limits(InteractiveConsole& console, [[maybe_unused]] cons
     console.WriteFormatLine("Map Elements: %zu/%d", tileElementCount, MAX_TILE_ELEMENTS);
     console.WriteFormatLine("Banners: %d/%zu", bannerCount, MAX_BANNERS);
     console.WriteFormatLine("Rides: %d/%d", rideCount, MAX_RIDES);
-    console.WriteFormatLine("Staff: %d/%d", staffCount, STAFF_MAX_COUNT);
```
revert. The staff count is still valid but the max is max_entities

duncanspumpkin commented on src/openrct2/config/Config.cpp:
```diff
@@ -475,7 +475,7 @@ namespace Config
     }
 
-    static void ReadWriteNotificationsChunk(IIniWriter* writer)
+    static void WriteNotificationsChunk(IIniWriter* writer)
```
This is still not fully reverted. Should drop the Chunk part of the change.

duncanspumpkin commented on src/openrct2-ui/windows/StaffList.cpp:
```diff
@@ -566,11 +566,7 @@ class StaffListWindow final : public Window
-                    continue;
-                }
-                if (!peep->IsLocationInPatrol(footpathCoords))
+                if (!peep->HasPatrolArea() || peep->IsLocationInPatrol(footpathCoords))
```
IsLocationInPatrol has been inverted. Causes #14824

Gymnasiast commented on src/openrct2/Context.cpp:
```diff
@@ -495,7 +495,6 @@ namespace OpenRCT2
 #endif
             }
 
-            gCurrentTicks = 0;
```
Needs to be reverted, I think.

ZehMatt replied:
GameState::InitAll takes care of this.

Gymnasiast commented on src/openrct2/ride/transport/Monorail.cpp:
```diff
@@ -460,7 +460,6 @@ static void paint_monorail_station(
 {
     uint32_t imageId;
     const StationObject* stationObject = nullptr;
-
```
revert

Gymnasiast commented on src/openrct2/ride/thrill/TopSpin.cpp:
```diff
@@ -44,7 +44,7 @@ static int8_t TopSpinSeatPositionOffset[] = {
 static void top_spin_paint_vehicle(
     paint_session* session, int32_t al, int32_t cl, const Ride* ride, uint8_t direction, int32_t height,
-    const TrackElement& tileElement)
+    const TrackElement& trackElement)
```
revert

Gymnasiast commented on src/openrct2/object/ImageTable.cpp:
```diff
@@ -214,7 +237,7 @@ std::vector<std::unique_ptr<ImageTable::RequiredImage>> ImageTable::LoadObjectIm
         if (placeHoldersAdded > 0)
         {
-            std::string msg = "Adding " + std::to_string(placeHoldersAdded) + " placeholders";
+            std::string msg = "PAdding " + std::to_string(placeHoldersAdded) + " placeholders";
```
revert

Gymnasiast commented on src/openrct2/localisation/StringIds.h:
```diff
@@ -3896,6 +3893,16 @@ enum
+    STR_TILE_INSPECTOR_FOOTPATH_RAILINGS_NAME = 6447,
+
+    STR_UNSUPPORTED_OBJECT_FORMAT = 6448,
+
```
STR_AT_LEAST_ONE_PATH_OBJECT_MUST_BE_SELECTED is no longer used and should be removed as well.

ZehMatt commented on src/openrct2/GameState.cpp:
```diff
@@ -58,6 +58,7 @@ void GameState::InitAll(int32_t mapSize)
     gInMapInitCode = true;
     gCurrentTicks = 0;
 
+    gCurrentTicks = 0;
```
This one can be be deleted.

Gymnasiast commented on src/openrct2/ride/Ride.h:
```diff
@@ -232,7 +233,7 @@ struct Ride
 
-private:
+public: // private:
     uint8_t min_max_cars_per_train;
```
This variable needs to be either split so that it can actually properly save the two, or removed altogether.

duncanspumpkin commented on src/openrct2/core/IStream.cpp:
```diff
@@ -12,9 +12,6 @@
 #include "Memory.hpp"
 #include "String.hpp"
-
-#include <vector>
```
revert this as everyone seems to enjoy having it here.

janisozaur commented on src/openrct2/ParkFile.cpp:
```diff
@@ -0,0 +1,4594 @@
+        cs.ReadWrite(entity.AssignedStaffType);
+        cs.ReadWrite(entity.MechanicTimeSinceCall);
+        cs.ReadWrite(entity.HireDate);
+        cs.ReadWrite(entity.StaffId);
```
Is this field still present?

janisozaur replied:
```
../src/openrct2/ParkFile.cpp:2010:29: error: 'struct Staff' has no member named 'StaffId'; did you mean 'Staff'?
         cs.ReadWrite(entity.StaffId);
                             ^~~~~~~
                             Staff
```

ZehMatt replied:
No, can be removed.

Gymnasiast commented on src/openrct2/paint/tile_element/Paint.Path.cpp:
```diff
@@ -1083,7 +1083,8 @@ void path_paint_box_support(
             { boundBoxOffset, height + boundingBoxZOffset });
 
-        if (!pathElement.IsQueue() && !pathElement.ShouldDrawPathOverSupports())
+        // TODO: Revert this when path import works correctly.
```
```suggestion
```

Gymnasiast commented on src/openrct2/paint/tile_element/Paint.Path.cpp:
```diff
@@ -1223,7 +1224,8 @@ void path_paint_pole_support(
             { boundBoxOffset, height + boundingBoxZOffset });
 
-        if (pathElement.IsQueue() || pathElement.ShouldDrawPathOverSupports())
+        // TODO: Revert this when path import works correctly.
```
```suggestion
```

duncanspumpkin commented on docs/save-format.md:
```diff
@@ -0,0 +1,255 @@
+
+Animations:
+
+Staff:
```
remove

duncanspumpkin commented on docs/save-format.md:
```diff
@@ -0,0 +1,255 @@
+  station-index:            uint8
+  colour:                   uint8
+  flags:                    uint8           (station / on-ride)
+  (spare: 5)
```
out of date

duncanspumpkin commented on src/openrct2-ui/windows/EditorObjectSelection.cpp:
```diff
@@ -101,19 +101,21 @@ static constexpr const ObjectPageDesc ObjectSelectionPages[] = {
-    // { STR_OBJECT_SELECTION_FOOTPATH_SURFACES,           SPR_TAB_SCENERY_PATHS,      false },
-    // { STR_OBJECT_SELECTION_FOOTPATH_RAILINGS,           SPR_G2_PATH_RAILINGS_TAB,   false },
+    // Dummy place holder for string objects
+    { STR_NONE,                   static_cast<uint32_t>(SPR_NONE),                  false },
```
remove cast?

duncanspumpkin commented on src/openrct2-ui/windows/EditorObjectSelection.cpp:
```diff
@@ -148,7 +150,7 @@ validate_global_widx(WC_EDITOR_OBJECT_SELECTION, WIDX_TAB_1);
 static std::vector<rct_widget> _window_editor_object_selection_widgets = {
     WINDOW_SHIM(WINDOW_TITLE, WW, WH),
-    MakeWidget({  0, 43}, {600, 357}, WindowWidgetType::Resize,       WindowColour::Secondary                                                                  ),
+    MakeWidget({  0, 43}, {WW, 357}, WindowWidgetType::Resize,       WindowColour::Secondary                                                                  ),
```
realign

duncanspumpkin commented on src/openrct2-ui/windows/EditorObjectSelection.cpp:
```diff
@@ -1371,26 +1368,27 @@ static void window_editor_object_selection_manage_tracks()
                 if (loadedObject == nullptr)
                 {
-                    log_error("Failed to load entry %.8s", entry->name);
+                    log_error("Failed to load entry %s", std::string(descriptor.GetName()).c_str());
```
Unsure if this is safe. Lifetime of std::string

ZehMatt replied:
Lifetime here ends after the call

Broxzier replied:
The `std::string` is a temporary and is valid until the end of the full expression, which is after `log_error` gets called, so this is perfectly safe.

duncanspumpkin commented on src/openrct2-ui/windows/Footpath.cpp:
```diff
@@ -749,6 +773,39 @@ static void window_footpath_show_footpath_types_dialog(rct_window* w, rct_widget
         gDropdownDefaultIndex = static_cast<int32_t>(*defaultIndex);
 }
 
+static void window_footpath_show_railings_types_dialog(rct_window* w, rct_widget* widget)
```
All new functions to be TitleCase

duncanspumpkin commented on src/openrct2-ui/windows/TileInspector.cpp:
```diff
@@ -1870,10 +1818,38 @@ static void window_tile_inspector_paint(rct_window* w, rct_drawpixelinfo* dpi)
-                auto ft = Formatter();
-                ft.Add<rct_string_id>(tileElement->AsPath()->GetSurfaceDescriptor()->Name);
-                DrawTextBasic(dpi, screenCoords, STR_TILE_INSPECTOR_PATH_NAME, ft, { w->colours[1] });
+                auto pathEl = tileElement->AsPath();
```
auto* for all this case

duncanspumpkin commented on src/openrct2/ParkFile.cpp:
```diff
@@ -0,0 +1,4593 @@
+    namespace ParkFileChunkType
+    {
+        // clang-format off
+//        constexpr uint32_t RESERVED_0       = 0x00;
```
realig

duncanspumpkin commented on src/openrct2/management/Research.cpp:
```diff
@@ -718,18 +724,17 @@ void research_remove_flags()
 }
 
-void research_fix()
+static void ResearchRemoveNullItems(std::vector<ResearchItem>& items)
```
Replace with `std::erase_if`

duncanspumpkin replied:
hmm erase_if is c++20. Will use a std::remove_if, erase.

duncanspumpkin commented on src/openrct2/world/Entrance.cpp:
```diff
@@ -217,6 +217,23 @@ void fix_park_entrance_locations(void)
         gParkEntrances.end());
 }
 
+void UpdateParkEntranceLocations()
```
Rename to `ResetParkEntranceLocations` or something else. Its not really updating them its setting them on first use. Should perhaps be moved to the parkfile as its only used by that.

duncanspumpkin commented on src/openrct2/world/Scenery.cpp:
```diff
@@ -394,3 +395,41 @@ void RestrictAllMiscScenery()
     }
 }
+
+ObjectType GetObjectTypeFromSceneryType(uint8_t type)
```
These are only used by the parkfile and then only for legacy interop. Suggest moving them back.

janisozaur reviewed (approved):
Done the usual test of compiling with latest upstream versions of compilers
